opm/input/eclipse/Schedule/Action/ActionParser.cpp

```diff
@@ -17,324 +17,583 @@
+#include <cstring>              // std::strlen()
+#include <memory>
 #include <stdexcept>
 #include <string>
 #include <utility>
 #include <vector>
 
 #include <fmt/format.h>
 
-namespace Opm::Action {
-
-Parser::Parser(const std::vector<std::string>& tokens_arg)
-    : tokens(tokens_arg)
-{}
-
-TokenType Parser::get_type(const std::string& arg)
+namespace {
+
+/// Classify ACTIONX condition sub expression according to summary vector
+/// category.
+///
+/// Mostly needed to aid condition evaluation later.
+///
+/// \param[in] arg Condition sub expression.  Expected to be the name of a
+/// summary vector.
+///
+/// \return Summary function category of \p arg.
+Opm::Action::FuncType functionType(const std::string& arg)
 {
-    std::string lower_arg = arg;
-    std::for_each(lower_arg.begin(),
-                  lower_arg.end(),
-                  [](char& c) {
-                      c = std::tolower(static_cast<unsigned char>(c));
-                  });
+    using FuncType = Opm::Action::FuncType;
 
-    if (lower_arg == "and") {
-        return TokenType::op_and;
+    if ((arg == "YEAR") || (arg == "DAY")) {
+        return FuncType::time;
     }
 
-    if (lower_arg == "or") {
-        return TokenType::op_or;
-    }
+    if (arg == "MNTH") { return FuncType::time_month; }
 
-    if (lower_arg == "(") {
-        return TokenType::open_paren;
-    }
+    using Cat = Opm::SummaryConfigNode::Category;
 
-    if (lower_arg == ")") {
-        return TokenType::close_paren;
+    switch (Opm::parseKeywordCategory(arg)) {
+    case Cat::Aquifer:    return FuncType::aquifer;
+    case Cat::Well:       return FuncType::well;
+    case Cat::Group:      return FuncType::group;
+    case Cat::Connection: return FuncType::well_connection;
+    case Cat::Region:     return FuncType::region;
+    case Cat::Block:      return FuncType::block;
+    case Cat::Segment:    return FuncType::well_segment;
+    default:              return FuncType::none;
     }
+}
 
-    if ((lower_arg == ">") || (lower_arg == ".gt.")) {
-        return TokenType::op_gt;
-    }
+/// Convert string to lower case characters.
+///
+/// \param[in] arg Input string.  Assumed to contain characters in the basic
+/// execution character set only.
+///
+/// \return Lower case copy of input string \p arg.
+std::string makeLowercase(const std::string& arg)
+{
+    std::string lower_arg = arg;
 
-    if ((lower_arg == ">=") || (lower_arg == ".ge.")) {
-        return TokenType::op_ge;
-    }
+    std::for_each(lower_arg.begin(), lower_arg.end(),
+                  [](char& c) {
+                      c = std::tolower(static_cast<unsigned char>(c));
+                  });
 
-    if ((lower_arg == "<") || (lower_arg == ".lt.")) {
-        return TokenType::op_lt;
-    }
+    return lower_arg;
+}
 
-    if ((lower_arg == "<=") || (lower_arg == ".le.")) {
-        return TokenType::op_le;
+/// Convert sequence of action condition textual tokens into an expression
+/// tree suitable for evaulation.
```

Comment:
evaluation

Reply:
> evaluation

Thanks!  I've pushed an update to that effect.